src/Parser.php

```diff
@@ -164,23 +166,23 @@ public function parse(string $uri): array
             return array_merge(self::URI_COMPONENTS, $simple_uri[$uri]);
         }
 
-        if (strlen($uri) !== strcspn($uri, self::INVALID_URI_CHARS)) {
+        if (preg_match(self::INVALID_CHARS_REGEXP, $uri)) {
             throw Exception::createFromInvalidCharacters($uri);
         }
 
         //if the first character is a known URI delimiter parsing can be simplified
         $first_char = $uri[0];
 
         //The URI is made of the fragment only
-        if ('#' == $first_char) {
+        if ('#' === $first_char) {
```

Comment:
while I understand the use of `===` in this case that make no much difference because `$first_char` will always be a `string`

Reply:
This is rather a generic code clean-up than a performance change. `==` usually shouldn't be used, at it can lead to subtle bugs.

Reply:
agreed, it's more a generic code clean up than a performance change. `===` is already used in critical place... but I have no objection to make it a requirement in the codebase